File: backend/src/models/funnel.py
"""Pydantic contract for the funnel endpoint — the server-owned board columns.

``GET /funnel`` returns the ordered board stages the SPA renders as columns; no stage name is
hardcoded in the SPA (DEV-PLAN slice 3 acceptance #9). Projected from :class:`core.funnel.Stage`.
"""

from __future__ import annotations

from pydantic import BaseModel


class FunnelStage(BaseModel):
    """One board column: a pipeline stage's short_name, label, order, and terminal flag."""

    short_name: str
    label: str
    sort_order: int
    is_terminal: bool

File: frontend/src/api/funnel.ts
import { useQuery, type UseQueryResult } from '@tanstack/react-query';
import { useApi } from './client';

// Mirrors backend models/funnel.py. The board columns come from the server (acceptance #9); the
// SPA hardcodes no stage name.

export interface FunnelStage {
  short_name: string;
  label: string;
  sort_order: number;
  is_terminal: boolean;
}

/** The server-owned board columns, in display order. Rarely changes, so it is cached indefinitely
 *  for the session; a full reload picks up any catalog change. */
export function useFunnel(): UseQueryResult<FunnelStage[]> {
  const api = useApi();
  return useQuery({
    queryKey: ['funnel'],
    queryFn: async () => (await api<{ stages: FunnelStage[] }>('/funnel')).stages,
    staleTime: Infinity,
  });
}